### PR TITLE
feat: apply resource filters on coming from service details to traces page

### PR DESCRIPTION
**Pull request by ahmadshaheer**
### Summary
 Actions in the serviceDetail view that take to the trace explorer page should have the resource attributes applied by default. Eg, selected host.name should translate to host.name = 0201ca93c5ae in the trace explorer page

#### Related Issues / PR's
close https://github.com/SigNoz/signoz/issues/2865

#### Screenshots

NA

### CODE REVIEW

**Comment by ankitnayan** [comment]
https://github.com/SigNoz/signoz/issues/2865#issuecomment-2325139565

File: frontend/src/container/MetricsApplication/Tabs/util.ts [hunk 1]
```diff
@@ -90,12 +90,15 @@ export function onGraphClickHandler(
...
+export const handleNonInQueryRange = (
+	tags: TagFilterItem[],
+	isTrace = false,
+): TagFilterItem[] =>
 	tags.map((tag) => {
 		if (tag.op === 'Not IN') {
 			return {
 				...tag,
-				op: 'NIN',
+				op: isTrace ? 'nin' : 'NIN',
```
Comment (srikanthccv) [inline frontend/src/container/MetricsApplication/Tabs/util.ts:101]: Why is this needed? The `NIN` should work

File: frontend/src/hooks/useResourceAttribute/utils.ts [hunk 2]
```diff
@@ -93,6 +93,22 @@ export const resourceAttributesToTagFilterItems = (
...
+		id: `${res.id}`,
+		key: {
+			key: convertMetricKeyToTrace(res.tagKey),
+			isColumn: false,
+			type: MetricsType.Resource,
+			dataType: DataTypes.String,
+			id: `${convertMetricKeyToTrace(res.tagKey)}--string--resource--true`,
+		},
+		op: `${res.operator}`,
+		value: `${res.tagValue}`.split(','),
```
Comment (srikanthccv) [inline frontend/src/hooks/useResourceAttribute/utils.ts:110]: What happens if there is a `,` in the value of the filter?

File: frontend/src/container/MetricsApplication/Tabs/util.ts (hunk 1 above)
Comment (ahmadshaheer) [reply frontend/src/container/MetricsApplication/Tabs/util.ts:101]: I tried with NIN, it works properly on going to traces from service details page, but was getting this error on subsequently refreshing the traces page in response of query_range.

```
{
  "status": "error",
  "errorType": "bad_data",
  "error": "multiple values [] are not allowed for operator `` for key `signoz.collector.id NIN a1b2c3d4-e5f6-4a5b-9c8d-7e6f5a4b3c2d`"
}
```
Note: signoz.collector.id in the above snippet has been changed to a random UUID.

Comment (ahmadshaheer) [reply frontend/src/container/MetricsApplication/Tabs/util.ts:101]: I guess the above issue (i.e. getting error in response of `query_range` on subsequent refresh) is a separate issue and can be addressed separately

Comment (ahmadshaheer) [reply frontend/src/container/MetricsApplication/Tabs/util.ts:101]: also, if I manually apply the same filter i.e. signoz.collector.id... using QB, it sets the operator as `nin`

File: frontend/src/hooks/useResourceAttribute/utils.ts (hunk 2 above)
Comment (ahmadshaheer) [reply frontend/src/hooks/useResourceAttribute/utils.ts:110]: @srikanthccv,

here the `res.tagValue` comes as an array. 

I got the snippet (`value: `${res.tagValue}`.split(','),`) from this [line](https://github.com/SigNoz/signoz/blob/b96b44444fa9009b49e13d57527c1a4a12676688/frontend/src/hooks/useResourceAttribute/utils.ts#L110), so I thought it would be necessary based on the format of filter in QB. 

When I tested now, we get the same result even if we don't have `.split(',')`. 

Should I remove it?

Comment (ahmadshaheer) [reply frontend/src/hooks/useResourceAttribute/utils.ts:110]: @srikanthccv, it's going to cause unintended behavior i.e. splitting the filter value. 

I had gotten it from this [line](https://github.com/SigNoz/signoz/blob/b96b44444fa9009b49e13d57527c1a4a12676688/frontend/src/hooks/useResourceAttribute/utils.ts#L93). I thought it would be necessary for the format of QB. but I tested by removing the `.split(',')`, and it works fine.

Removed it.

File: frontend/src/hooks/useResourceAttribute/utils.ts [hunk 3]
```diff
@@ -93,6 +93,22 @@ export const resourceAttributesToTagFilterItems = (
...
+	queries: IResourceAttribute[],
+): TagFilterItem[] =>
+	queries.map((res) => ({
+		id: `${res.id}`,
+		key: {
+			key: convertMetricKeyToTrace(res.tagKey),
+			isColumn: false,
+			type: MetricsType.Resource,
+			dataType: DataTypes.String,
+			id: `${convertMetricKeyToTrace(res.tagKey)}--string--resource--true`,
```
Comment (SagarRajput-7) [inline frontend/src/hooks/useResourceAttribute/utils.ts:107]: better to also spread `res` at the end to not miss any current and future key-value like `isJSON`

Comment (srikanthccv) [reply frontend/src/hooks/useResourceAttribute/utils.ts:107]: This is not required in the request payload. Keep it if it's required on the frontend otherwise, remove it.

Comment (ahmadshaheer) [reply frontend/src/hooks/useResourceAttribute/utils.ts:107]: @SagarRajput-7, here we get res in the following format:
```
{
    "id": "1cc16351",
    "tagKey": "resource_host_name",
    "operator": "Not IN",
    "tagValue": [
        "abc3421ab"
    ]
}
```

I guess we don't need to spread it. Thoughts?

**Comment by ankitnayan** [comment]
@ahmadshaheer please record a video of `environment` and `host.name` being propagated to trace explorer page from the serviceList -> serviceDetail page. I am not able to see it work

**Comment by ahmadshaheer** [comment]
> @ahmadshaheer please record a video of `environment` and `host.name` being propagated to trace explorer page from the serviceList -> serviceDetail page. I am not able to see it work

Sure.  https://signoz-team.slack.com/archives/C02TJ466H8U/p1725890834952129

**Review by SagarRajput-7 — APPROVED** [review]

**Review by YounixM — APPROVED** [review]

**Review by srikanthccv — APPROVED** [review]
Didn't test the functionality myself. Approving to unblock merge.

**Review by ankitnayan — APPROVED** [review]

**Comment by ahmadshaheer** [comment]
> > [#5827 (comment)](https://github.com/SigNoz/signoz/pull/5827#discussion_r1746584077)
> 
> @ahmadshaheer please create an issue for this and merge this PR by fixing the build and resolving conv

Sure, @ankitnayan.

https://github.com/SigNoz/signoz/issues/5913
